{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Cytoscape Jupyter Widget Demo\n",
    "\n",
    "![](https://cytoscape.org/images/logo/cy3logoOrange.svg)\n",
    "\n",
    "![](cyjupyter1.png)\n",
    "\n",
    "By Keiichiro ONO (kono at ucsd dot edu)\n",
    "\n",
    "Updated on 10/11/2018\n",
    "\n",
    "\n",
    "## Introduction\n",
    "\n",
    "This is a demo notebook for Cytoscape Jupyter Widget: a network visualizer for Jupyter Notebook.  It's a very simple graph visualizer using [Cytoscape.js](http://js.cytoscape.org).  You can interactively visualize graphs (networks) in CX or Cytoscape.js format.\n",
    "\n",
    "\n",
    "## Requirments\n",
    "\n",
    "This notebook was tested with the following:\n",
    "\n",
    "* cyjupyter v0.2.0\n",
    "* Jupyter Notebook v5.6\n",
    "* NetworkX v2.2\n",
    "\n",
    "\n",
    "### Official Web Site and Code Repository\n",
    "\n",
    "* https://github.com/idekerlab/cytoscape-jupyter-widget \n",
    "\n",
    "\n",
    "## Install Widget\n",
    "\n",
    "### From PyPi\n",
    "\n",
    "```bash\n",
    "pip install cyjupyter\n",
    "jupyter nbextension enable --py --sys-prefix cyjupyter\n",
    "```\n",
    "\n",
    "\n",
    "### From Source\n",
    "\n",
    "```bash\n",
    "git cloneã€€https://github.com/idekerlab/cytoscape-jupyter-widget.git\n",
    "cd cytoscape-jupyter-widget\n",
    "./build.sh\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## How to Use the Widget\n",
    "\n",
    "### 1. Import Widget"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cyjupyter import Cytoscape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Prepare Data\n",
    "\n",
    "This widget supports graph data in CX or Cytoscape.js JSON (cyjs).\n",
    "\n",
    "#### Cytoscape.js JSON\n",
    "\n",
    "Cytoscape.js JSON is a simple format for graph data.  The following code generates Cytoscape.js data as a standard Python dictionary.  Yu can find the complete documentation about this format [here](http://js.cytoscape.org/#notation/elements-json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# A Cytoscape.js network with 4 nodes and 5 edges represented as Python Dict\n",
    "\n",
    "minimal_cyjs_network = {\n",
    "    'elements': {\n",
    "        'nodes': [\n",
    "            {'data' : {'id': 'node 1'}},\n",
    "            {'data' : {'id': 'node 2'}},\n",
    "            {'data' : {'id': 'node 3'}},\n",
    "            {'data' : {'id': 'node 4'}}\n",
    "        ],\n",
    "        'edges': [\n",
    "            { 'data': {'id': 'edge1','source': 'node 1','target': 'node 2'}},\n",
    "            { 'data': {'id': 'edge2','source': 'node 1','target': 'node 3'}},\n",
    "            { 'data': {'id': 'edge3','source': 'node 2','target': 'node 3'}},\n",
    "            { 'data': {'id': 'edge4','source': 'node 2','target': 'node 4'}},\n",
    "            { 'data': {'id': 'edge5','source': 'node 3','target': 'node 4'}}\n",
    "        ]\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CX JSON\n",
    "\n",
    "CX is a JSON format used for network data transfer in the Cytoscape ecosystem.  You can download any network from [NDEx](http://www.ndexbio.org/) in CX format.  The following example uses NDEx Python client to download the network in CX.  You can change the _uuid_ to try any networks in NDEx."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "starting download\n",
      "Response code from NDEx:  200\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[{'numberVerification': [{'longNumber': 281474976710655}]},\n",
       " {'metaData': [{'name': 'nodeAttributes',\n",
       "    'elementCount': 25,\n",
       "    'version': '1.0'},\n",
       "   {'name': 'cyHiddenAttributes', 'elementCount': 3, 'version': '1.0'},\n",
       "   {'name': 'nodes', 'elementCount': 25, 'idCounter': 24, 'version': '1.0'},\n",
       "   {'name': 'networkAttributes', 'elementCount': 11, 'version': '1.0'},\n",
       "   {'name': 'cyTableColumn', 'elementCount': 30, 'version': '1.0'},\n",
       "   {'name': 'cartesianLayout', 'elementCount': 25, 'version': '1.0'},\n",
       "   {'name': 'edgeAttributes', 'elementCount': 488, 'version': '1.0'},\n",
       "   {'name': 'edges', 'elementCount': 69, 'idCounter': 68, 'version': '1.0'},\n",
       "   {'name': 'cyVisualProperties', 'elementCount': 3, 'version': '1.0'},\n",
       "   {'name': 'cartesianCoordinates', 'elementCount': 1}]},\n",
       " {'nodeAttributes': [{'po': 24, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 23, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 22, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 21, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 20, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 19, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 18, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 17, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 16, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 15, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 14, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 13, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 12, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 11, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 10, 'n': 'TYPE', 'v': 'phenotype'},\n",
       "   {'po': 9, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 8, 'n': 'TYPE', 'v': 'phenotype'},\n",
       "   {'po': 7, 'n': 'TYPE', 'v': 'phenotype'},\n",
       "   {'po': 6, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 5, 'n': 'TYPE', 'v': 'complex'},\n",
       "   {'po': 4, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 3, 'n': 'TYPE', 'v': 'complex'},\n",
       "   {'po': 2, 'n': 'TYPE', 'v': 'protein'},\n",
       "   {'po': 1, 'n': 'TYPE', 'v': 'complex'},\n",
       "   {'po': 0, 'n': 'TYPE', 'v': 'protein'}]},\n",
       " {'cyHiddenAttributes': [{'n': 'NDEx UUID',\n",
       "    'v': '5e507821-e221-11e8-aaa6-0ac135e8bacf'},\n",
       "   {'n': 'layoutAlgorithm', 'v': 'Prefuse Force Directed Layout'},\n",
       "   {'n': 'NDEx Modification Timestamp', 'v': '2018-11-06 16:09:17.893'}]},\n",
       " {'nodes': [{'@id': 24, 'n': 'FADD', 'r': 'uniprot:Q13158'},\n",
       "   {'@id': 23, 'n': 'CASP8', 'r': 'uniprot:Q14790'},\n",
       "   {'@id': 22, 'n': 'TRADD', 'r': 'uniprot:Q15628'},\n",
       "   {'@id': 21, 'n': 'TNF', 'r': 'uniprot:P01375'},\n",
       "   {'@id': 20, 'n': 'TNFRSF1A', 'r': 'uniprot:P19438'},\n",
       "   {'@id': 19, 'n': 'BAG4', 'r': 'uniprot:O95429'},\n",
       "   {'@id': 18, 'n': 'MAP2K3', 'r': 'uniprot:P46734'},\n",
       "   {'@id': 17, 'n': 'MAPK14', 'r': 'uniprot:Q16539'},\n",
       "   {'@id': 16, 'n': 'MAP2K7', 'r': 'uniprot:O14733'},\n",
       "   {'@id': 15, 'n': 'MAPK8', 'r': 'uniprot:P45983'},\n",
       "   {'@id': 14, 'n': 'MAP2K4', 'r': 'uniprot:P45985'},\n",
       "   {'@id': 13, 'n': 'MAP3K1', 'r': 'uniprot:Q13233'},\n",
       "   {'@id': 12, 'n': 'MAP3K5', 'r': 'uniprot:Q99683'},\n",
       "   {'@id': 11, 'n': 'TRAF2', 'r': 'uniprot:Q12933'},\n",
       "   {'@id': 10, 'n': 'Proliferation', 'r': 'signor:SIGNOR-PH4'},\n",
       "   {'@id': 9, 'n': 'JUN', 'r': 'uniprot:P05412'},\n",
       "   {'@id': 8, 'n': 'Inflammation', 'r': 'signor:SIGNOR-PH12'},\n",
       "   {'@id': 7, 'n': 'Apoptosis', 'r': 'signor:SIGNOR-PH2'},\n",
       "   {'@id': 6, 'n': 'CASP3', 'r': 'uniprot:P42574'},\n",
       "   {'@id': 5, 'n': 'NfKb-p65/p50', 'r': 'signor:SIGNOR-C13'},\n",
       "   {'@id': 4, 'n': 'NFKBIA', 'r': 'uniprot:P25963'},\n",
       "   {'@id': 3, 'n': 'IKK-complex', 'r': 'signor:SIGNOR-C14'},\n",
       "   {'@id': 2, 'n': 'RIPK1', 'r': 'uniprot:Q13546'},\n",
       "   {'@id': 1, 'n': 'BAK/BAX', 'r': 'signor:SIGNOR-C96'},\n",
       "   {'@id': 0, 'n': 'BID', 'r': 'uniprot:P55957'}]},\n",
       " {'networkAttributes': [{'n': 'rightsHolder', 'v': 'Prof. Gianni Cesareni'},\n",
       "   {'n': 'organism', 'v': 'Human, 9606, Homo sapiens'},\n",
       "   {'n': 'description',\n",
       "    'v': 'The Tumor necrosis factor modulates a vast array of cell functions including inflammation, apoptosis, cell proliferation, immunity and differentiation. TNF alpha is mainly produced by macrophages. <p><br/></p><h6><b>Node Legend:</b><br/>Light green oval &gt; Protein/Protein Family<br/>Dark green round rectangle &gt; Complex<br/>Orange octagon &gt; Chemical<br/>Purple octagon &gt; Small molecule<br/>White rectangles &gt; Phenotype<br/>Light blue diamond &gt; Stimulus</h6><h6><b>Edge Legend:</b><br/>Solid &gt; Direct interaction<br/>Dashed &gt; Indirect or Unknown interaction<br/>Blue &gt; Up-regulation<br/>Red &gt; Down-regulation<br/>Black &gt; Form complex or Unknown</h6>'},\n",
       "   {'n': 'version', 'v': '05-Oct-2018'},\n",
       "   {'n': '@context',\n",
       "    'v': '{\"pubmed\":\"http://identifiers.org/pubmed/\",\"chebi\":\"http://identifiers.org/chebi/CHEBI:\",\"signor\":\"http://signor.uniroma2.it/relation_result.php?id\\\\u003d\",\"BTO\":\"http://identifiers.org/bto/BTO:\",\"uniprot\":\"http://identifiers.org/uniprot/\",\"CID\":\"http://identifiers.org/pubchem.compound/\",\"SID\":\"http://identifiers.org/pubchem.substance/\"}'},\n",
       "   {'n': 'labels', 'v': ['SIGNOR-TA'], 'd': 'list_of_string'},\n",
       "   {'n': 'reference',\n",
       "    'v': '<div>Perfetto L., <i>et al.</i></div><div><b>SIGNOR: a database of causal relationships between biological entities</b><i>.</i></div><div>Nucleic Acids Res. 2016 Jan 4;44(D1):D548-54</div><div><span><a href=\"https://doi.org/10.1093/nar/gkv1048\" target=\"\">doi: 10.1093/nar/gkv1048</a></span></div>'},\n",
       "   {'n': 'rights',\n",
       "    'v': 'Attribution-ShareAlike 4.0 International (CC BY-SA 4.0)'},\n",
       "   {'n': 'name', 'v': 'TNF alpha -test'},\n",
       "   {'n': 'networkType', 'v': 'Signalling Pathway'},\n",
       "   {'n': 'dataSource',\n",
       "    'v': 'https://signor.uniroma2.it/pathway_browser.php?organism=&pathway_list=SIGNOR-TA'}]},\n",
       " {'cyTableColumn': [{'applies_to': 'node_table', 'n': 'shared name'},\n",
       "   {'applies_to': 'node_table', 'n': 'name'},\n",
       "   {'applies_to': 'node_table', 'n': 'TYPE'},\n",
       "   {'applies_to': 'edge_table', 'n': 'shared name'},\n",
       "   {'applies_to': 'edge_table', 'n': 'shared interaction'},\n",
       "   {'applies_to': 'edge_table', 'n': 'name'},\n",
       "   {'applies_to': 'edge_table', 'n': 'interaction'},\n",
       "   {'applies_to': 'edge_table', 'n': 'MECHANISM'},\n",
       "   {'applies_to': 'edge_table', 'n': 'citation', 'd': 'list_of_string'},\n",
       "   {'applies_to': 'edge_table', 'n': 'SENTENCE'},\n",
       "   {'applies_to': 'edge_table', 'n': 'ANNOTATOR'},\n",
       "   {'applies_to': 'edge_table', 'n': 'DIRECT'},\n",
       "   {'applies_to': 'edge_table', 'n': 'CELL_DATA', 'd': 'list_of_string'},\n",
       "   {'applies_to': 'edge_table', 'n': 'SEQUENCE'},\n",
       "   {'applies_to': 'edge_table', 'n': 'RESIDUE'},\n",
       "   {'applies_to': 'edge_table', 'n': 'TISSUE_DATA', 'd': 'list_of_string'},\n",
       "   {'applies_to': 'network_table', 'n': 'shared name'},\n",
       "   {'applies_to': 'network_table', 'n': 'name'},\n",
       "   {'applies_to': 'network_table', 'n': 'rightsHolder'},\n",
       "   {'applies_to': 'network_table', 'n': 'organism'},\n",
       "   {'applies_to': 'network_table', 'n': 'author'},\n",
       "   {'applies_to': 'network_table', 'n': 'description'},\n",
       "   {'applies_to': 'network_table', 'n': 'version'},\n",
       "   {'applies_to': 'network_table', 'n': 'labels', 'd': 'list_of_string'},\n",
       "   {'applies_to': 'network_table', 'n': 'reference'},\n",
       "   {'applies_to': 'network_table', 'n': 'rights'},\n",
       "   {'applies_to': 'network_table', 'n': 'networkType'},\n",
       "   {'applies_to': 'network_table', 'n': 'dataSource'},\n",
       "   {'applies_to': 'network_table', 'n': '@context'},\n",
       "   {'applies_to': 'network_table',\n",
       "    'n': '__Annotations',\n",
       "    'd': 'list_of_string'}]},\n",
       " {'cartesianLayout': [{'node': 2,\n",
       "    'x': 13.438410505056368,\n",
       "    'y': -92.36262287139891},\n",
       "   {'node': 10, 'x': 299.8452250802517, 'y': 262.2887748336792},\n",
       "   {'node': 18, 'x': -79.78212477326394, 'y': 164.1419929122925},\n",
       "   {'node': 3, 'x': 155.05915330290793, 'y': -38.14643634796141},\n",
       "   {'node': 11, 'x': -46.994771257638945, 'y': -74.53745807647704},\n",
       "   {'node': 19, 'x': -118.58995558381082, 'y': -395.217862739563},\n",
       "   {'node': 4, 'x': 236.53010056853293, 'y': 37.32848583221437},\n",
       "   {'node': 12, 'x': -42.787991777658476, 'y': 47.62117229461671},\n",
       "   {'node': 20, 'x': -31.725957170724882, 'y': -313.4855629348755},\n",
       "   {'node': 5, 'x': 286.47861741423606, 'y': 145.9913581466675},\n",
       "   {'node': 13, 'x': 30.022963269948946, 'y': 9.439730033874525},\n",
       "   {'node': 21, 'x': 6.1004702889919145, 'y': -411.1585976028442},\n",
       "   {'node': 6, 'x': -220.8538410818577, 'y': -16.7880836868286},\n",
       "   {'node': 14, 'x': -6.13652540802957, 'y': 130.9327062225342},\n",
       "   {'node': 22, 'x': -16.149815813303007, 'y': -193.38787616729735},\n",
       "   {'node': 0, 'x': -219.90282179474832, 'y': 55.56287609100343},\n",
       "   {'node': 7, 'x': -345.6977436697483, 'y': -2.7074791336059434},\n",
       "   {'node': 15, 'x': 81.55658601164816, 'y': 196.9569877243042},\n",
       "   {'node': 23, 'x': -104.77023817658426, 'y': -37.86192096710204},\n",
       "   {'node': 8, 'x': 403.2334391915798, 'y': 170.10058055877687},\n",
       "   {'node': 16, 'x': 67.05509413719176, 'y': 93.48945461273195},\n",
       "   {'node': 24, 'x': -94.74573256134988, 'y': -140.96197284698485},\n",
       "   {'node': 1, 'x': -330.8101704275608, 'y': 105.75412403106691},\n",
       "   {'node': 9, 'x': 192.5515544259548, 'y': 224.5061820602417},\n",
       "   {'node': 17, 'x': -112.92392470002176, 'y': 72.50144802093507}]},\n",
       " {'edgeAttributes': [{'po': 64, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 64,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:22890322'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 64,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Rip1 is required for the formation of a rip1/fadd/caspase-8 complex that drives caspase-8 activation, cleavage of bid into tbid, mitochondrial outer membrane permeabilization, full activation of caspase-3 and caspase-dependent apoptosis. Tweak induces assembly of a death-signaling complex containing rip1, fadd, and caspase-8'},\n",
       "   {'po': 64, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 64, 'n': 'name', 'v': 'FADD (up-regulates activity) RIPK1'},\n",
       "   {'po': 64, 'n': 'TISSUE_DATA', 'v': ['BTO:0000931'], 'd': 'list_of_string'},\n",
       "   {'po': 64, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 63, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 63,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:21525013'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 63,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Rip1 is required for the formation of a rip1/fadd/caspase-8 complex that drives caspase-8 activation, cleavage of bid into tbid, mitochondrial outer membrane permeabilization, full activation of caspase-3 and caspase-dependent apoptosis. Tweak induces assembly of a death-signaling complex containing rip1, fadd, and caspase-8'},\n",
       "   {'po': 63, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 63, 'n': 'name', 'v': 'FADD (up-regulates activity) RIPK1'},\n",
       "   {'po': 63, 'n': 'CELL_DATA', 'v': ['BTO:0002025'], 'd': 'list_of_string'},\n",
       "   {'po': 63, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 56, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 56,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:11717445'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 56,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Fadd recruits caspase-8 through homotypic interactions of death-effector domains (deds), leading to caspase-8 activation and apoptosis. In turn, fadd recruits the zymogen form of the apoptosis-initiating protease caspase-8, through homophilic interaction of death effector domains.'},\n",
       "   {'po': 56, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 56, 'n': 'name', 'v': 'FADD (up-regulates activity) CASP8'},\n",
       "   {'po': 56, 'n': 'CELL_DATA', 'v': ['BTO:0000782'], 'd': 'list_of_string'},\n",
       "   {'po': 56, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 65, 'n': 'MECHANISM', 'v': 'cleavage'},\n",
       "   {'po': 65,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:10521396'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 65,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'These results suggested that the aspartic acid at position 324 is the cleavage site of ripk1. In this study we found that receptor-interacting protein (ripk1) is cleaved by casp8 when cells undergo tnf-induced apoptosis. The cleavage of ripk1 abolished its nf-kb inducing ability.'},\n",
       "   {'po': 65, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 65, 'n': 'SEQUENCE', 'v': 'RMQSLQLdCVAVPSS'},\n",
       "   {'po': 65, 'n': 'name', 'v': 'CASP8 (down-regulates activity) RIPK1'},\n",
       "   {'po': 65,\n",
       "    'n': 'CELL_DATA',\n",
       "    'v': ['BTO:0000007', 'BTO:0000093', 'BTO:0000567'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 65, 'n': 'RESIDUE', 'v': 'Asp324'},\n",
       "   {'po': 65, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 59, 'n': 'MECHANISM', 'v': 'cleavage'},\n",
       "   {'po': 59,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:16964285'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 59,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Casp8 induces apoptosis by directly activating casp3.'},\n",
       "   {'po': 59, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 59, 'n': 'name', 'v': 'CASP8 (up-regulates activity) CASP3'},\n",
       "   {'po': 59, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 59, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 58, 'n': 'MECHANISM', 'v': 'cleavage'},\n",
       "   {'po': 58, 'n': 'citation', 'v': ['pubmed:9727492'], 'd': 'list_of_string'},\n",
       "   {'po': 58,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Caspase-8 cleaves bid at aspartic acid residue 60 (asp60) cleavage of bid by casp8 releases its potent proapoptotic activity'},\n",
       "   {'po': 58, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 58, 'n': 'SEQUENCE', 'v': 'GYDELQTdGNRSSHS'},\n",
       "   {'po': 58, 'n': 'name', 'v': 'CASP8 (up-regulates activity) BID'},\n",
       "   {'po': 58, 'n': 'CELL_DATA', 'v': ['BTO:0000093'], 'd': 'list_of_string'},\n",
       "   {'po': 58, 'n': 'RESIDUE', 'v': 'Asp60'},\n",
       "   {'po': 58, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 68, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 68, 'n': 'citation', 'v': ['pubmed:8612133'], 'd': 'list_of_string'},\n",
       "   {'po': 68,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'We show that tradd interacts strongly with rip;rip is a serinethreonine kinase that is recruited by tradd to tnfr1 in a tnf-dependent process.'},\n",
       "   {'po': 68, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 68, 'n': 'name', 'v': 'TRADD (up-regulates activity) RIPK1'},\n",
       "   {'po': 68, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 68, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 67, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 67,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:18545270'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 67,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'It also associates with death domain-containing adaptor proteins, such as tnf-receptor-associated death domain (tradd) that are required for caspase-8 activation and apoptosis.'},\n",
       "   {'po': 67, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 67, 'n': 'name', 'v': 'TRADD (up-regulates activity) RIPK1'},\n",
       "   {'po': 67, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 62, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 62,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:18545270'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 62, 'n': 'SENTENCE', 'v': 'Tradd recruits fadd'},\n",
       "   {'po': 62, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 62, 'n': 'name', 'v': 'TRADD (up-regulates activity) FADD'},\n",
       "   {'po': 62, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 61, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 61, 'n': 'citation', 'v': ['pubmed:8565075'], 'd': 'list_of_string'},\n",
       "   {'po': 61,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The strong interaction between tradd and fadd occurs via their death domains.'},\n",
       "   {'po': 61, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 61, 'n': 'name', 'v': 'TRADD (up-regulates activity) FADD'},\n",
       "   {'po': 61, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 61, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 53, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 53,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:18621737'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 53,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The high affinity of the tradd-traf2 interaction is required for efficient suppression of apoptosis upon stimulation of the tumor necrosis factor receptor1 (tnfr1), tnf-receptor-associated death domain (tradd) provides a scaffold for the assembly of complex i at the plasma membrane by binding receptor interacting protein 1 (rip1), tnfreceptor- associated factor 2 ,traf2 these results provide evidence that tradd can serve as an adaptor protein and recruit traf1, traf2, or both to tnfrsf1a. The demonstration that tradd interacts with traf2 and fadd, and can recruit both to tnfrsf1a, suggested that traf2 and fadd may be involved in tnfrsf1a tradd-mediated signaling. That these interactions define two distinct signaling pathways emanating from tradd (figure 9) is supported by the ability of traf2 and fadd to activate nf-kb and induce apoptosis, respectively.'},\n",
       "   {'po': 53, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 53, 'n': 'name', 'v': 'TRADD (up-regulates activity) TRAF2'},\n",
       "   {'po': 53, 'n': 'CELL_DATA', 'v': ['BTO:0000459'], 'd': 'list_of_string'},\n",
       "   {'po': 53, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 52, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 52,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:14585074'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 52,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Tradd mediates recruitment of the traf2 adaptor protein'},\n",
       "   {'po': 52, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 52, 'n': 'name', 'v': 'TRADD (up-regulates activity) TRAF2'},\n",
       "   {'po': 52, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 51, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 51, 'n': 'citation', 'v': ['pubmed:8702708'], 'd': 'list_of_string'},\n",
       "   {'po': 51,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The high affinity of the tradd-traf2 interaction is required for efficient suppression of apoptosis upon stimulation of the tumor necrosis factor receptor1 (tnfr1), tnf-receptor-associated death domain (tradd) provides a scaffold for the assembly of complex i at the plasma membrane by binding receptor interacting protein 1 (rip1), tnfreceptor- associated factor 2 ,traf2 these results provide evidence that tradd can serve as an adaptor protein and recruit traf1, traf2, or both to tnfrsf1a. The demonstration that tradd interacts with traf2 and fadd, and can recruit both to tnfrsf1a, suggested that traf2 and fadd may be involved in tnfrsf1a tradd-mediated signaling. That these interactions define two distinct signaling pathways emanating from tradd (figure 9) is supported by the ability of traf2 and fadd to activate nf-kb and induce apoptosis, respectively.'},\n",
       "   {'po': 51, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 51, 'n': 'name', 'v': 'TRADD (up-regulates activity) TRAF2'},\n",
       "   {'po': 51, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 51, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 48, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 48,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:10634209'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 48,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'TNF-induced apoptosis is mediated primarily through the activation of type I receptors'},\n",
       "   {'po': 48, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 48, 'n': 'name', 'v': 'TNF (up-regulates activity) TNFRSF1A'},\n",
       "   {'po': 48, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 47, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 47,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:23070005'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 47,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'For tnfr1, the cytokine tnfalfa binds to the receptor and triggers its trimerization the trimeric tnf induces tnfrsf1a aggregation, which is stabilized by homotypic interactions of the death domain of tnfrsf1ain a solid phase binding assay rstnfr beta was found to bind tnf alpha with high affinity comparable with the membrane-bound full-length receptor.'},\n",
       "   {'po': 47, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 47, 'n': 'name', 'v': 'TNF (up-regulates activity) TNFRSF1A'},\n",
       "   {'po': 47, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 46, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 46,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:14965271'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 46, 'n': 'SENTENCE', 'v': 'TNFalpha activates TNF-R1'},\n",
       "   {'po': 46, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 46, 'n': 'name', 'v': 'TNF (up-regulates activity) TNFRSF1A'},\n",
       "   {'po': 46, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 45, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 45,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:11502070'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 45,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Binding of tnf to the extracellular domain of tnfrsf1a leads to homotrimerization.'},\n",
       "   {'po': 45, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 45, 'n': 'name', 'v': 'TNF (up-regulates activity) TNFRSF1A'},\n",
       "   {'po': 45, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 50, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 50, 'n': 'citation', 'v': ['pubmed:7758105'], 'd': 'list_of_string'},\n",
       "   {'po': 50,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'We have identified a novel 34 kda protein, designated tradd, that specifically interacts with an intracellular domain of tnfr1 tradd interacts with the death domain of tnfrsf1a to initiate distinct signaling cascades for two of the most important biological activities of tnf, nf-kb activation and programmed cell death tradd, a novel protein that specifically interacts with the death domain of tnfr1 and activates signaling pathways for both of these activities when overexpressed.'},\n",
       "   {'po': 50, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 50, 'n': 'name', 'v': 'TNFRSF1A (up-regulates activity) TRADD'},\n",
       "   {'po': 50, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 50, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 49, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 49,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:11502070'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 49,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The death domain of tnfrsf1a provides a novel molecular interface that interacts specifically with the death domain of tradd.'},\n",
       "   {'po': 49, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 49, 'n': 'name', 'v': 'TNFRSF1A (up-regulates activity) TRADD'},\n",
       "   {'po': 49, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 44, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 44,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:12748303'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 44,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'It was suggested that the silencer of death domains (SODD) protein constitutively associates intracellularly with TNFR1 and inhibits the recruitment of cytoplasmic signaling proteins to TNFR1 to prevent spontaneous aggregation of the cytoplasmic death domains of TNFR1 molecules that are juxtaposed in the absence of ligand stimulation'},\n",
       "   {'po': 44, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 44, 'n': 'name', 'v': 'BAG4 (down-regulates activity) TNFRSF1A'},\n",
       "   {'po': 44, 'n': 'CELL_DATA', 'v': ['BTO:0000801'], 'd': 'list_of_string'},\n",
       "   {'po': 44, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 33, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 33, 'n': 'citation', 'v': ['pubmed:7839144'], 'd': 'list_of_string'},\n",
       "   {'po': 33,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Two human MAP kinase kinases (MKK3 and MKK4) were cloned that phosphorylate and activate p38 MAP kinase.'},\n",
       "   {'po': 33, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 33, 'n': 'name', 'v': 'MAP2K3 (up-regulates activity) MAPK14'},\n",
       "   {'po': 33, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 32, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 32, 'n': 'citation', 'v': ['pubmed:8622669'], 'd': 'list_of_string'},\n",
       "   {'po': 32,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Mkk3 is a protein kinase that phosphorylates and activates p38 map kinase but does not phosphorylate the related jnk or erk map kinases.'},\n",
       "   {'po': 32, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 32, 'n': 'name', 'v': 'MAP2K3 (up-regulates activity) MAPK14'},\n",
       "   {'po': 32, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 31, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 31, 'n': 'citation', 'v': ['pubmed:7535770'], 'd': 'list_of_string'},\n",
       "   {'po': 31,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Recently, two map kinase kinases (mkk3 and mkk4) that activate p38 map kinase have been identified. the mechanism of p38 activation is mediated by dual phosphorylation on thr-180 and tyr-182.'},\n",
       "   {'po': 31, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 31, 'n': 'SEQUENCE', 'v': 'TDDEMTGyVATRWYR'},\n",
       "   {'po': 31, 'n': 'name', 'v': 'MAP2K3 (up-regulates activity) MAPK14'},\n",
       "   {'po': 31,\n",
       "    'n': 'CELL_DATA',\n",
       "    'v': ['BTO:0000130', 'BTO:0000801', 'BTO:0000876'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 31, 'n': 'RESIDUE', 'v': 'Tyr182'},\n",
       "   {'po': 31, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 30, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 30, 'n': 'citation', 'v': ['pubmed:8622669'], 'd': 'list_of_string'},\n",
       "   {'po': 30,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Mkk3 is a protein kinase that phosphorylates and activates p38 map kinase but does not phosphorylate the related jnk or erk map kinases.'},\n",
       "   {'po': 30, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 30, 'n': 'SEQUENCE', 'v': 'TDDEMTGyVATRWYR'},\n",
       "   {'po': 30, 'n': 'name', 'v': 'MAP2K3 (up-regulates activity) MAPK14'},\n",
       "   {'po': 30, 'n': 'RESIDUE', 'v': 'Tyr182'},\n",
       "   {'po': 30, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 29, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 29, 'n': 'citation', 'v': ['pubmed:8622669'], 'd': 'list_of_string'},\n",
       "   {'po': 29,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Mkk3 is a protein kinase that phosphorylates and activates p38 map kinase but does not phosphorylate the related jnk or erk map kinases.'},\n",
       "   {'po': 29, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 29, 'n': 'SEQUENCE', 'v': 'RHTDDEMtGYVATRW'},\n",
       "   {'po': 29, 'n': 'name', 'v': 'MAP2K3 (up-regulates activity) MAPK14'},\n",
       "   {'po': 29, 'n': 'RESIDUE', 'v': 'Thr180'},\n",
       "   {'po': 29, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 60, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 60,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:14970175'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 60,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Consequently, p38-mapk can directly phosphorylate and inhibit the activities of caspase-8 and caspase-3 and thereby hinder neutrophil apoptosis, and, in so doing, regulate the inflammatory response.'},\n",
       "   {'po': 60, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 60, 'n': 'SEQUENCE', 'v': 'FRGDRCRsLTGKPKL'},\n",
       "   {'po': 60, 'n': 'name', 'v': 'MAPK14 (down-regulates) CASP3'},\n",
       "   {'po': 60, 'n': 'CELL_DATA', 'v': ['BTO:0000130'], 'd': 'list_of_string'},\n",
       "   {'po': 60, 'n': 'RESIDUE', 'v': 'Ser150'},\n",
       "   {'po': 60, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 57, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 57,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:14970175'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 57,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'P38-mapk can directly phosphorylate and inhibit the activities of caspase-8'},\n",
       "   {'po': 57, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 57, 'n': 'SEQUENCE', 'v': 'FTGLKCPsLAGKPKV'},\n",
       "   {'po': 57, 'n': 'name', 'v': 'MAPK14 (down-regulates) CASP8'},\n",
       "   {'po': 57, 'n': 'CELL_DATA', 'v': ['BTO:0000130'], 'd': 'list_of_string'},\n",
       "   {'po': 57, 'n': 'RESIDUE', 'v': 'Ser347'},\n",
       "   {'po': 57, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 23, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 23, 'n': 'citation', 'v': ['pubmed:9312068'], 'd': 'list_of_string'},\n",
       "   {'po': 23,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Jnk is activated by jnk-activating kinase 1 (jnkk1), a dual specificity protein kinase that phosphorylates jnk on threonine 183 and tyrosine 185 residues.'},\n",
       "   {'po': 23, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 23, 'n': 'SEQUENCE', 'v': 'AGTSFMMtPYVVTRY'},\n",
       "   {'po': 23, 'n': 'name', 'v': 'MAP2K7 (up-regulates) MAPK8'},\n",
       "   {'po': 23, 'n': 'RESIDUE', 'v': 'Thr183'},\n",
       "   {'po': 23, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 22, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 22,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:11062067'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 22,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Jnk full activation requires the phosphorylation of a threonine and a tyrosine residue in a thr-pro-tyr motif, which can be catalysed by the protein kinases mitogen-activated protein kinase kinase (mkk)4 and mkk7. Mkk7 shows a striking preference for the threonine residue (thr-183).'},\n",
       "   {'po': 22, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 22, 'n': 'SEQUENCE', 'v': 'AGTSFMMtPYVVTRY'},\n",
       "   {'po': 22, 'n': 'name', 'v': 'MAP2K7 (up-regulates) MAPK8'},\n",
       "   {'po': 22, 'n': 'RESIDUE', 'v': 'Thr183'},\n",
       "   {'po': 22, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 21, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 21, 'n': 'citation', 'v': ['pubmed:9312068'], 'd': 'list_of_string'},\n",
       "   {'po': 21,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Jnk is activated by jnk-activating kinase 1 (jnkk1), a dual specificity protein kinase that phosphorylates jnk on threonine 183 and tyrosine 185 residues.'},\n",
       "   {'po': 21, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 21, 'n': 'SEQUENCE', 'v': 'TSFMMTPyVVTRYYR'},\n",
       "   {'po': 21, 'n': 'name', 'v': 'MAP2K7 (up-regulates) MAPK8'},\n",
       "   {'po': 21, 'n': 'RESIDUE', 'v': 'Tyr185'},\n",
       "   {'po': 21, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 25, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 25, 'n': 'citation', 'v': ['pubmed:8137421'], 'd': 'list_of_string'},\n",
       "   {'po': 25,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': ' JNK1 binds to the c-Jun transactivation domain and phosphorylates it on Ser-63 and Ser-73. The effect on AP-1 transcriptional activity results, in part, from enhanced phosphorylation of the c-Jun NH2-terminal activation domain.'},\n",
       "   {'po': 25, 'n': 'ANNOTATOR', 'v': 'miannu'},\n",
       "   {'po': 25, 'n': 'SEQUENCE', 'v': 'VGLLKLAsPELERLI'},\n",
       "   {'po': 25, 'n': 'name', 'v': 'MAPK8 (up-regulates activity) JUN'},\n",
       "   {'po': 25, 'n': 'CELL_DATA', 'v': ['BTO:0000298'], 'd': 'list_of_string'},\n",
       "   {'po': 25, 'n': 'RESIDUE', 'v': 'Ser73'},\n",
       "   {'po': 25, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 24, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 24, 'n': 'citation', 'v': ['pubmed:8137421'], 'd': 'list_of_string'},\n",
       "   {'po': 24,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The jnk-mediated phosphorylation of both ser63 and ser73 within the transactivation domain of c-jun potentiates its transcriptional activity.'},\n",
       "   {'po': 24, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 24, 'n': 'SEQUENCE', 'v': 'KNSDLLTsPDVGLLK'},\n",
       "   {'po': 24, 'n': 'name', 'v': 'MAPK8 (up-regulates activity) JUN'},\n",
       "   {'po': 24, 'n': 'CELL_DATA', 'v': ['BTO:0004055'], 'd': 'list_of_string'},\n",
       "   {'po': 24, 'n': 'RESIDUE', 'v': 'Ser63'},\n",
       "   {'po': 24, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 28, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 28, 'n': 'citation', 'v': ['pubmed:7535770'], 'd': 'list_of_string'},\n",
       "   {'po': 28,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Recently, two MAP kinase kinases (MKK3 and MKK4) that activate p38 MAP kinase have been identified. The mechanism of p38 activation is mediated by dual phosphorylation on Thr-180 and Tyr-182'},\n",
       "   {'po': 28, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 28, 'n': 'SEQUENCE', 'v': 'RHTDDEMtGYVATRW'},\n",
       "   {'po': 28, 'n': 'name', 'v': 'MAP2K4 (up-regulates activity) MAPK14'},\n",
       "   {'po': 28, 'n': 'RESIDUE', 'v': 'Thr180'},\n",
       "   {'po': 28, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 27, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 27, 'n': 'citation', 'v': ['pubmed:7839144'], 'd': 'list_of_string'},\n",
       "   {'po': 27,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Two human MAP kinase kinases (MKK3 and MKK4) were cloned that phosphorylate and activate p38 MAP kinase.'},\n",
       "   {'po': 27, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 27, 'n': 'name', 'v': 'MAP2K4 (up-regulates activity) MAPK14'},\n",
       "   {'po': 27, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 20, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 20, 'n': 'citation', 'v': ['pubmed:9724739'], 'd': 'list_of_string'},\n",
       "   {'po': 20,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'MKK4/7, in turn, phosphorylates JNK on residues 183 and 185 (17\\x83\\x82\\x80\\x9320). Activated JNK phosphorylates its substrates, c-Jun, ATF2, ELK1, and p53'},\n",
       "   {'po': 20, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 20, 'n': 'SEQUENCE', 'v': 'AGTSFMMtPYVVTRY'},\n",
       "   {'po': 20, 'n': 'name', 'v': 'MAP2K4 (up-regulates activity) MAPK8'},\n",
       "   {'po': 20, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 20, 'n': 'RESIDUE', 'v': 'Thr183'},\n",
       "   {'po': 20, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 19, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 19, 'n': 'citation', 'v': ['pubmed:9724739'], 'd': 'list_of_string'},\n",
       "   {'po': 19,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'MKK4/7, in turn, phosphorylates JNK on residues 183 and 185 (17\\x83\\x82\\x80\\x9320). Activated JNK phosphorylates its substrates, c-Jun, ATF2, ELK1, and p53'},\n",
       "   {'po': 19, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 19, 'n': 'SEQUENCE', 'v': 'TSFMMTPyVVTRYYR'},\n",
       "   {'po': 19, 'n': 'name', 'v': 'MAP2K4 (up-regulates activity) MAPK8'},\n",
       "   {'po': 19, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 19, 'n': 'RESIDUE', 'v': 'Tyr185'},\n",
       "   {'po': 19, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 41, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 41, 'n': 'citation', 'v': ['pubmed:9712898'], 'd': 'list_of_string'},\n",
       "   {'po': 41,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The gck-ctd-mekk1 interaction is sufficiently stable to support mekk1 s phosphorylation of its substrate, sek1'},\n",
       "   {'po': 41, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 41, 'n': 'SEQUENCE', 'v': 'LVDSIAKtRDAGCRP'},\n",
       "   {'po': 41, 'n': 'name', 'v': 'MAP3K1 (up-regulates activity) MAP2K4'},\n",
       "   {'po': 41, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 41, 'n': 'RESIDUE', 'v': 'Thr261'},\n",
       "   {'po': 41, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 40, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 40, 'n': 'citation', 'v': ['pubmed:9712898'], 'd': 'list_of_string'},\n",
       "   {'po': 40,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The gck-ctd-mekk1 interaction is sufficiently stable to support mekk1 s phosphorylation of its substrate, SEK1'},\n",
       "   {'po': 40, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 40, 'n': 'name', 'v': 'MAP3K1 (up-regulates activity) MAP2K4'},\n",
       "   {'po': 40, 'n': 'CELL_DATA', 'v': ['BTO:0000007'], 'd': 'list_of_string'},\n",
       "   {'po': 40, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 37, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 37, 'n': 'citation', 'v': ['pubmed:9312068'], 'd': 'list_of_string'},\n",
       "   {'po': 37,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Here we show that jnkk2, a novel member of the map kinase kinase family, was phosphorylated and activated by mekk1'},\n",
       "   {'po': 37, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 37, 'n': 'SEQUENCE', 'v': 'LVDSKAKtRSAGCAA'},\n",
       "   {'po': 37, 'n': 'name', 'v': 'MAP3K1 (up-regulates) MAP2K7'},\n",
       "   {'po': 37, 'n': 'RESIDUE', 'v': 'Thr275'},\n",
       "   {'po': 37, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 36, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 36, 'n': 'citation', 'v': ['pubmed:9312068'], 'd': 'list_of_string'},\n",
       "   {'po': 36,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Here we show that jnkk2, a novel member of the map kinase kinase family, was phosphorylated and activated by mekk1'},\n",
       "   {'po': 36, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 36, 'n': 'SEQUENCE', 'v': 'ISGRLVDsKAKTRSA'},\n",
       "   {'po': 36, 'n': 'name', 'v': 'MAP3K1 (up-regulates) MAP2K7'},\n",
       "   {'po': 36, 'n': 'RESIDUE', 'v': 'Ser271'},\n",
       "   {'po': 36, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 43, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 43, 'n': 'citation', 'v': ['pubmed:8974401'], 'd': 'list_of_string'},\n",
       "   {'po': 43,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'A MAP kinase kinase kinase (MAPKKK), termed ASK1, was identified that activated two different subgroups of MAP kinase kinases (MAPKK), SEK1 (or MKK4) and MKK3/MAPKK6 (or MKK6), which in turn activated stress-activated protein kinase (SAPK, also known as JNK; c-Jun amino-terminal kinase) and p38 subgroups of MAP kinases, respectively.'},\n",
       "   {'po': 43, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 43, 'n': 'name', 'v': 'MAP3K5 (up-regulates activity) MAP2K4'},\n",
       "   {'po': 43, 'n': 'CELL_DATA', 'v': ['BTO:0000298'], 'd': 'list_of_string'},\n",
       "   {'po': 43, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 42, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 42, 'n': 'citation', 'v': ['pubmed:8974401'], 'd': 'list_of_string'},\n",
       "   {'po': 42,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'A map kinase kinase kinase (mapkkk), termed ask1, was identified that activated two different subs of map kinase kinases (mapkk), sek1 (or mkk4) and mkk3/mapkk6 (or mkk6), which in turn activated stress-activated protein kinase (sapk, also known as jnk;c-jun amino-terminal kinase)'},\n",
       "   {'po': 42, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 42, 'n': 'SEQUENCE', 'v': 'LVDSIAKtRDAGCRP'},\n",
       "   {'po': 42, 'n': 'name', 'v': 'MAP3K5 (up-regulates activity) MAP2K4'},\n",
       "   {'po': 42, 'n': 'RESIDUE', 'v': 'Thr261'},\n",
       "   {'po': 42, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 39, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 39,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:19920149'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 39,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Ask1 is a member of a mapkkk family and functions as an upstream kinase engaged in c-jun nh2-terminal kinase (jnk)/p38 signaling via the phosphorylation and activation of mapkks, such as mkk3, -4, -6, and -7'},\n",
       "   {'po': 39, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 39, 'n': 'name', 'v': 'MAP3K5 (up-regulates) MAP2K7'},\n",
       "   {'po': 39, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 38, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 38,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:11959862'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 38, 'n': 'SENTENCE', 'v': 'Activation of mkk7 by ask1'},\n",
       "   {'po': 38, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 38, 'n': 'name', 'v': 'MAP3K5 (up-regulates) MAP2K7'},\n",
       "   {'po': 38, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 35, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 35,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:19920149'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 35,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Ask1 is a member of a mapkkk family and functions as an upstream kinase engaged in c-jun nh2-terminal kinase (jnk)/p38 signaling via the phosphorylation and activation of mapkks, such as mkk3, -4, -6, and -7'},\n",
       "   {'po': 35, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 35, 'n': 'SEQUENCE', 'v': 'ISGYLVDsVAKTMDA'},\n",
       "   {'po': 35, 'n': 'name', 'v': 'MAP3K5 (up-regulates activity) MAP2K3'},\n",
       "   {'po': 35, 'n': 'RESIDUE', 'v': 'Ser218'},\n",
       "   {'po': 35, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 34, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 34, 'n': 'citation', 'v': ['pubmed:8974401'], 'd': 'list_of_string'},\n",
       "   {'po': 34,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'ASK1 activated SEK1 and MKK3 up to 7.0- and 11.8-fold, respectively'},\n",
       "   {'po': 34, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 34, 'n': 'name', 'v': 'MAP3K5 (up-regulates activity) MAP2K3'},\n",
       "   {'po': 34, 'n': 'CELL_DATA', 'v': ['BTO:0004055'], 'd': 'list_of_string'},\n",
       "   {'po': 34, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 66, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 66,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:21232017'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 66,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Rip1 is known to directly interact with traf2'},\n",
       "   {'po': 66, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 66, 'n': 'name', 'v': 'TRAF2 (up-regulates activity) RIPK1'},\n",
       "   {'po': 66,\n",
       "    'n': 'CELL_DATA',\n",
       "    'v': ['BTO:0002572', 'BTO:0000801'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 66, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 18, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 18,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:10346818'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 18,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Oligomerization of the traf2 effector domain results in specific binding to mekk1, a protein kinase capable of jnk, p38, and ikk activation'},\n",
       "   {'po': 18, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 18, 'n': 'name', 'v': 'TRAF2 (up-regulates) MAP3K1'},\n",
       "   {'po': 18, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 17, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 17,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:18635759'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 17,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Traf2, ubc13, and ikkgamma were required for complex assembly and activation of mekk1 and mapk cascades.'},\n",
       "   {'po': 17, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 17, 'n': 'name', 'v': 'TRAF2 (up-regulates) MAP3K1'},\n",
       "   {'po': 17, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 10, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 10,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:10688666'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 10,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Tnf receptor (tnfr) associated factor 2 (traf2), an adapter protein that couples tnfrs to the sapks and p38s, can activate ask1 in vivo and can interact in vivo with the amino- and carboxyl-terminal noncatalytic domains of the ask1 polypeptide'},\n",
       "   {'po': 10, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 10, 'n': 'name', 'v': 'TRAF2 (up-regulates activity) MAP3K5'},\n",
       "   {'po': 10, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 9, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 9, 'n': 'citation', 'v': ['pubmed:9774977'], 'd': 'list_of_string'},\n",
       "   {'po': 9, 'n': 'SENTENCE', 'v': 'Traf2 is a strong activator of ask1'},\n",
       "   {'po': 9, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 9, 'n': 'name', 'v': 'TRAF2 (up-regulates activity) MAP3K5'},\n",
       "   {'po': 9, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 7, 'n': 'citation', 'v': ['pubmed:9878062'], 'd': 'list_of_string'},\n",
       "   {'po': 7,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Functional data suggest that c-Jun is not merely a target for activation by many of the extracellular stimuli, but that it plays a role in mediating the cellular response. In the case of growth control, three lines of evidence suggest that the transcription factor AP-1, which is composed of Fos\\x96Jun and Jun\\x96Jun dimers, mediates cell proliferation in response to external growth signals in the form of peptide growth factors.'},\n",
       "   {'po': 7, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 7, 'n': 'name', 'v': 'JUN (up-regulates) Proliferation'},\n",
       "   {'po': 7, 'n': 'DIRECT', 'v': 'NO'},\n",
       "   {'po': 4, 'n': 'citation', 'v': ['pubmed:14585074'], 'd': 'list_of_string'},\n",
       "   {'po': 4,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Caspase-3 is responsible for apoptosis execution'},\n",
       "   {'po': 4, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 4, 'n': 'name', 'v': 'CASP3 (up-regulates) Apoptosis'},\n",
       "   {'po': 4, 'n': 'DIRECT', 'v': 'NO'},\n",
       "   {'po': 26, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 26,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:18174238'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 26,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Chromatin immunoprecipitation (chip) analysis confirmed the serum-induced recruitment of jund to the promoter in vivo and showed that the presence of jund was dependent on the presence of p65 and p50, indicating a protein-protein-dependent mechanism of jund recruitment'},\n",
       "   {'po': 26, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 26, 'n': 'name', 'v': 'NfKb-p65/p50 (up-regulates) JUN'},\n",
       "   {'po': 26, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 8, 'n': 'citation', 'v': ['pubmed:11359934'], 'd': 'list_of_string'},\n",
       "   {'po': 8,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The nuclear factor-kappaB (NF-kappaB) family of transcription factors has been shown to regulate proliferation in several cell types.'},\n",
       "   {'po': 8, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 8, 'n': 'name', 'v': 'NfKb-p65/p50 (up-regulates) Proliferation'},\n",
       "   {'po': 8, 'n': 'DIRECT', 'v': 'NO'},\n",
       "   {'po': 6, 'n': 'citation', 'v': ['pubmed:20457564'], 'd': 'list_of_string'},\n",
       "   {'po': 6,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The nuclear factor NF-kappaB pathway has long been considered a prototypical proinflammatory signaling pathway, largely based on the role of NF-kappaB in the expression of proinflammatory genes including cytokines, chemokines, and adhesion molecules.'},\n",
       "   {'po': 6, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 6, 'n': 'name', 'v': 'NfKb-p65/p50 (up-regulates) Inflammation'},\n",
       "   {'po': 6, 'n': 'DIRECT', 'v': 'NO'},\n",
       "   {'po': 3, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 3, 'n': 'citation', 'v': ['pubmed:9914500'], 'd': 'list_of_string'},\n",
       "   {'po': 3,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'In nonstimulated cells, nf-kappab is present in the cytosol where it is complexed to its inhibitor ikappab  however, we found that only one of the activities, namely the ikk1/2 complex, exists as a pre-assembled kinase-substrate complex in which the ikks are directly or indirectly associated with several nf-kappab-related and ikappab-related proteins: rela, relb, crel, p100, p105, ikappa balpha, ikappa bbeta and ikappa bepsilon.'},\n",
       "   {'po': 3, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 3,\n",
       "    'n': 'name',\n",
       "    'v': 'NFKBIA (down-regulates activity) NfKb-p65/p50'},\n",
       "   {'po': 3, 'n': 'CELL_DATA', 'v': ['BTO:0000567'], 'd': 'list_of_string'},\n",
       "   {'po': 3, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 2, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 2, 'n': 'citation', 'v': ['pubmed:1340770'], 'd': 'list_of_string'},\n",
       "   {'po': 2,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Nf-kappa b is an inducible transcription factor comprised of a 50-kd (p50) and a 65-kd (p65) subunit. Induction of nf-kappa b activity, which is a critical event in many signal transduction pathways, involves release from a cytoplasmic inhibitory protein, i kappa b, followed by translocation of the active transcription factor complex into the nucleus. we demonstrate by in vitro and in vivo methods that the recently cloned i kappa b/mad-3 interacts with both the p50 and p65 subunits of nf-kappa b'},\n",
       "   {'po': 2, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 2, 'n': 'name', 'v': 'NFKBIA (down-regulates) NfKb-p65/p50'},\n",
       "   {'po': 2, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 14, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 14, 'n': 'citation', 'v': ['pubmed:9346241'], 'd': 'list_of_string'},\n",
       "   {'po': 14,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'We described the purification of a 900 kda protein kinase complex, the ikb kinase (ikk), that phosphorylates ikbalfa and ikbbeta at the sites that mediate their ubiquitination and degradation'},\n",
       "   {'po': 14, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 14, 'n': 'SEQUENCE', 'v': 'RHDSGLDsMKDEEYE'},\n",
       "   {'po': 14,\n",
       "    'n': 'name',\n",
       "    'v': 'IKK-complex (down-regulates quantity by destabilization) NFKBIA'},\n",
       "   {'po': 14, 'n': 'CELL_DATA', 'v': ['BTO:0000567'], 'd': 'list_of_string'},\n",
       "   {'po': 14, 'n': 'RESIDUE', 'v': 'Ser36'},\n",
       "   {'po': 14, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 13, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 13,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:21232017'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 13,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Tak1 become activated and then phosphorilates and activates ikk2 which in turn now phosphorylates ikba, marking it for k48-ubiquitination and proteasomal degradation'},\n",
       "   {'po': 13, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 13, 'n': 'name', 'v': 'IKK-complex (down-regulates) NFKBIA'},\n",
       "   {'po': 13, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 12, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 12, 'n': 'citation', 'v': ['pubmed:9346241'], 'd': 'list_of_string'},\n",
       "   {'po': 12,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'We described the purification of a 900 kda protein kinase complex, the ikb kinase (ikk), that phosphorylates ikbalfa and ikbbeta at the sites that mediate their ubiquitination and degradation'},\n",
       "   {'po': 12, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 12, 'n': 'SEQUENCE', 'v': 'RHDSGLDsMKDEEYE'},\n",
       "   {'po': 12, 'n': 'name', 'v': 'IKK-complex (down-regulates) NFKBIA'},\n",
       "   {'po': 12, 'n': 'RESIDUE', 'v': 'Ser36'},\n",
       "   {'po': 12, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 11, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 11, 'n': 'citation', 'v': ['pubmed:9346241'], 'd': 'list_of_string'},\n",
       "   {'po': 11,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'We described the purification of a 900 kda protein kinase complex, the ikb kinase (ikk), that phosphorylates ikbalfa and ikbbeta at the sites that mediate their ubiquitination and degradation'},\n",
       "   {'po': 11, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 11, 'n': 'SEQUENCE', 'v': 'LLDDRHDsGLDSMKD'},\n",
       "   {'po': 11, 'n': 'name', 'v': 'IKK-complex (down-regulates) NFKBIA'},\n",
       "   {'po': 11, 'n': 'RESIDUE', 'v': 'Ser32'},\n",
       "   {'po': 11, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 55, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 55,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:21525013'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 55,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Degradation of ciaps triggers the release of receptor interacting protein kinase (ripk1) from tnf receptor i (tnfr1) to form a caspase-8 activating complex'},\n",
       "   {'po': 55, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 55, 'n': 'name', 'v': 'RIPK1 (up-regulates activity) CASP8'},\n",
       "   {'po': 55, 'n': 'CELL_DATA', 'v': ['BTO:0002025'], 'd': 'list_of_string'},\n",
       "   {'po': 55, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 54, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 54,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:12887920'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 54,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Tradd and rip1 associate with fadd and caspase-8, forming a cytoplasmic complex'},\n",
       "   {'po': 54, 'n': 'ANNOTATOR', 'v': 'amattioni'},\n",
       "   {'po': 54, 'n': 'name', 'v': 'RIPK1 (up-regulates activity) CASP8'},\n",
       "   {'po': 54,\n",
       "    'n': 'CELL_DATA',\n",
       "    'v': ['BTO:0000007', 'BTO:0000459'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 54, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 16, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 16,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:11369754'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 16,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'These findings strongly suggest that rip phosphorylates mekk1 at ser-957 and ser-994.'},\n",
       "   {'po': 16, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 16, 'n': 'SEQUENCE', 'v': 'HSQCLNSsPLSHHSQ'},\n",
       "   {'po': 16, 'n': 'name', 'v': 'RIPK1 (up-regulates activity) MAP3K1'},\n",
       "   {'po': 16, 'n': 'CELL_DATA', 'v': ['BTO:0000661'], 'd': 'list_of_string'},\n",
       "   {'po': 16, 'n': 'RESIDUE', 'v': 'Ser970'},\n",
       "   {'po': 16, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 15, 'n': 'MECHANISM', 'v': 'phosphorylation'},\n",
       "   {'po': 15,\n",
       "    'n': 'citation',\n",
       "    'v': ['pubmed:11369754'],\n",
       "    'd': 'list_of_string'},\n",
       "   {'po': 15,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'These findings strongly suggest that rip phosphorylates mekk1 at ser-957 and ser-994.'},\n",
       "   {'po': 15, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 15, 'n': 'name', 'v': 'RIPK1 (up-regulates) MAP3K1'},\n",
       "   {'po': 15, 'n': 'CELL_DATA', 'v': ['BTO:0000661'], 'd': 'list_of_string'},\n",
       "   {'po': 15, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 1, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 1, 'n': 'citation', 'v': ['pubmed:10795740'], 'd': 'list_of_string'},\n",
       "   {'po': 1,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'We found that TNF-R1-mediated IKK activation requires both RIP and TRAF2 proteins. Although TRAF2 or RIP can be independently recruited to the TNF-R1 complex, neither one of them alone is capable of transducing the TNF signal that leads to IKK activation'},\n",
       "   {'po': 1, 'n': 'ANNOTATOR', 'v': 'gcesareni'},\n",
       "   {'po': 1, 'n': 'name', 'v': 'RIPK1 (up-regulates activity) IKK-complex'},\n",
       "   {'po': 1, 'n': 'CELL_DATA', 'v': ['BTO:0000452'], 'd': 'list_of_string'},\n",
       "   {'po': 1, 'n': 'DIRECT', 'v': 'YES'},\n",
       "   {'po': 5, 'n': 'citation', 'v': ['pubmed:23567751'], 'd': 'list_of_string'},\n",
       "   {'po': 5,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'The mitochondrial pathway of apoptosis proceeds when the outer mitochondrial membrane (OMM) is compromised by the pro-apoptotic BCL-2 family members, BAK and BAX. Once activated, BAK and BAX form proteolipid pores in the OMM leading to mitochondrial outer membrane permeabilization (MOMP), and the release of inner membrane space proteins, such as cytochrome c, which promotes caspase activation.'},\n",
       "   {'po': 5, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 5, 'n': 'name', 'v': 'BAK/BAX (up-regulates) Apoptosis'},\n",
       "   {'po': 5, 'n': 'DIRECT', 'v': 'NO'},\n",
       "   {'po': 0, 'n': 'MECHANISM', 'v': 'binding'},\n",
       "   {'po': 0, 'n': 'citation', 'v': ['pubmed:16243507'], 'd': 'list_of_string'},\n",
       "   {'po': 0,\n",
       "    'n': 'SENTENCE',\n",
       "    'v': 'Letai et al. [ 16] proposed that BH3-only proteins comprised both sensitisers, which only inactivate the pro-survival proteins, and activators, which directly engage Bax and Bak. In this model ( Figure 4a), the activators, proposed to include tBid and Bim, may normally be sequestered by the pro-survival proteins.'},\n",
       "   {'po': 0, 'n': 'ANNOTATOR', 'v': 'lperfetto'},\n",
       "   {'po': 0, 'n': 'name', 'v': 'BID (up-regulates activity) BAK/BAX'},\n",
       "   {'po': 0, 'n': 'DIRECT', 'v': 'YES'}]},\n",
       " {'edges': [{'@id': 64, 's': 24, 't': 2, 'i': 'up-regulates activity'},\n",
       "   {'@id': 63, 's': 24, 't': 2, 'i': 'up-regulates activity'},\n",
       "   {'@id': 56, 's': 24, 't': 23, 'i': 'up-regulates activity'},\n",
       "   {'@id': 65, 's': 23, 't': 2, 'i': 'down-regulates activity'},\n",
       "   {'@id': 59, 's': 23, 't': 6, 'i': 'up-regulates activity'},\n",
       "   {'@id': 58, 's': 23, 't': 0, 'i': 'up-regulates activity'},\n",
       "   {'@id': 68, 's': 22, 't': 2, 'i': 'up-regulates activity'},\n",
       "   {'@id': 67, 's': 22, 't': 2, 'i': 'up-regulates activity'},\n",
       "   {'@id': 62, 's': 22, 't': 24, 'i': 'up-regulates activity'},\n",
       "   {'@id': 61, 's': 22, 't': 24, 'i': 'up-regulates activity'},\n",
       "   {'@id': 53, 's': 22, 't': 11, 'i': 'up-regulates activity'},\n",
       "   {'@id': 52, 's': 22, 't': 11, 'i': 'up-regulates activity'},\n",
       "   {'@id': 51, 's': 22, 't': 11, 'i': 'up-regulates activity'},\n",
       "   {'@id': 48, 's': 21, 't': 20, 'i': 'up-regulates activity'},\n",
       "   {'@id': 47, 's': 21, 't': 20, 'i': 'up-regulates activity'},\n",
       "   {'@id': 46, 's': 21, 't': 20, 'i': 'up-regulates activity'},\n",
       "   {'@id': 45, 's': 21, 't': 20, 'i': 'up-regulates activity'},\n",
       "   {'@id': 50, 's': 20, 't': 22, 'i': 'up-regulates activity'},\n",
       "   {'@id': 49, 's': 20, 't': 22, 'i': 'up-regulates activity'},\n",
       "   {'@id': 44, 's': 19, 't': 20, 'i': 'down-regulates activity'},\n",
       "   {'@id': 33, 's': 18, 't': 17, 'i': 'up-regulates activity'},\n",
       "   {'@id': 32, 's': 18, 't': 17, 'i': 'up-regulates activity'},\n",
       "   {'@id': 31, 's': 18, 't': 17, 'i': 'up-regulates activity'},\n",
       "   {'@id': 30, 's': 18, 't': 17, 'i': 'up-regulates activity'},\n",
       "   {'@id': 29, 's': 18, 't': 17, 'i': 'up-regulates activity'},\n",
       "   {'@id': 60, 's': 17, 't': 6, 'i': 'down-regulates'},\n",
       "   {'@id': 57, 's': 17, 't': 23, 'i': 'down-regulates'},\n",
       "   {'@id': 23, 's': 16, 't': 15, 'i': 'up-regulates'},\n",
       "   {'@id': 22, 's': 16, 't': 15, 'i': 'up-regulates'},\n",
       "   {'@id': 21, 's': 16, 't': 15, 'i': 'up-regulates'},\n",
       "   {'@id': 25, 's': 15, 't': 9, 'i': 'up-regulates activity'},\n",
       "   {'@id': 24, 's': 15, 't': 9, 'i': 'up-regulates activity'},\n",
       "   {'@id': 28, 's': 14, 't': 17, 'i': 'up-regulates activity'},\n",
       "   {'@id': 27, 's': 14, 't': 17, 'i': 'up-regulates activity'},\n",
       "   {'@id': 20, 's': 14, 't': 15, 'i': 'up-regulates activity'},\n",
       "   {'@id': 19, 's': 14, 't': 15, 'i': 'up-regulates activity'},\n",
       "   {'@id': 41, 's': 13, 't': 14, 'i': 'up-regulates activity'},\n",
       "   {'@id': 40, 's': 13, 't': 14, 'i': 'up-regulates activity'},\n",
       "   {'@id': 37, 's': 13, 't': 16, 'i': 'up-regulates'},\n",
       "   {'@id': 36, 's': 13, 't': 16, 'i': 'up-regulates'},\n",
       "   {'@id': 43, 's': 12, 't': 14, 'i': 'up-regulates activity'},\n",
       "   {'@id': 42, 's': 12, 't': 14, 'i': 'up-regulates activity'},\n",
       "   {'@id': 39, 's': 12, 't': 16, 'i': 'up-regulates'},\n",
       "   {'@id': 38, 's': 12, 't': 16, 'i': 'up-regulates'},\n",
       "   {'@id': 35, 's': 12, 't': 18, 'i': 'up-regulates activity'},\n",
       "   {'@id': 34, 's': 12, 't': 18, 'i': 'up-regulates activity'},\n",
       "   {'@id': 66, 's': 11, 't': 2, 'i': 'up-regulates activity'},\n",
       "   {'@id': 18, 's': 11, 't': 13, 'i': 'up-regulates'},\n",
       "   {'@id': 17, 's': 11, 't': 13, 'i': 'up-regulates'},\n",
       "   {'@id': 10, 's': 11, 't': 12, 'i': 'up-regulates activity'},\n",
       "   {'@id': 9, 's': 11, 't': 12, 'i': 'up-regulates activity'},\n",
       "   {'@id': 7, 's': 9, 't': 10, 'i': 'up-regulates'},\n",
       "   {'@id': 4, 's': 6, 't': 7, 'i': 'up-regulates'},\n",
       "   {'@id': 26, 's': 5, 't': 9, 'i': 'up-regulates'},\n",
       "   {'@id': 8, 's': 5, 't': 10, 'i': 'up-regulates'},\n",
       "   {'@id': 6, 's': 5, 't': 8, 'i': 'up-regulates'},\n",
       "   {'@id': 3, 's': 4, 't': 5, 'i': 'down-regulates activity'},\n",
       "   {'@id': 2, 's': 4, 't': 5, 'i': 'down-regulates'},\n",
       "   {'@id': 14,\n",
       "    's': 3,\n",
       "    't': 4,\n",
       "    'i': 'down-regulates quantity by destabilization'},\n",
       "   {'@id': 13, 's': 3, 't': 4, 'i': 'down-regulates'},\n",
       "   {'@id': 12, 's': 3, 't': 4, 'i': 'down-regulates'},\n",
       "   {'@id': 11, 's': 3, 't': 4, 'i': 'down-regulates'},\n",
       "   {'@id': 55, 's': 2, 't': 23, 'i': 'up-regulates activity'},\n",
       "   {'@id': 54, 's': 2, 't': 23, 'i': 'up-regulates activity'},\n",
       "   {'@id': 16, 's': 2, 't': 13, 'i': 'up-regulates activity'},\n",
       "   {'@id': 15, 's': 2, 't': 13, 'i': 'up-regulates'},\n",
       "   {'@id': 1, 's': 2, 't': 3, 'i': 'up-regulates activity'},\n",
       "   {'@id': 5, 's': 1, 't': 7, 'i': 'up-regulates'},\n",
       "   {'@id': 0, 's': 0, 't': 1, 'i': 'up-regulates activity'}]},\n",
       " {'cyVisualProperties': [{'properties_of': 'network',\n",
       "    'properties': {'NETWORK_ANNOTATION_SELECTION': 'false',\n",
       "     'NETWORK_BACKGROUND_PAINT': '#FFFFFF',\n",
       "     'NETWORK_CENTER_X_LOCATION': '0.0',\n",
       "     'NETWORK_CENTER_Y_LOCATION': '0.0',\n",
       "     'NETWORK_CENTER_Z_LOCATION': '0.0',\n",
       "     'NETWORK_DEPTH': '0.0',\n",
       "     'NETWORK_EDGE_SELECTION': 'true',\n",
       "     'NETWORK_FORCE_HIGH_DETAIL': 'false',\n",
       "     'NETWORK_HEIGHT': '707.0',\n",
       "     'NETWORK_NODE_LABEL_SELECTION': 'false',\n",
       "     'NETWORK_NODE_SELECTION': 'true',\n",
       "     'NETWORK_SCALE_FACTOR': '1.0',\n",
       "     'NETWORK_SIZE': '550.0',\n",
       "     'NETWORK_WIDTH': '1141.0'}},\n",
       "   {'properties_of': 'nodes:default',\n",
       "    'properties': {'COMPOUND_NODE_PADDING': '10.0',\n",
       "     'COMPOUND_NODE_SHAPE': 'OCTAGON',\n",
       "     'NODE_BORDER_PAINT': '#009900',\n",
       "     'NODE_BORDER_STROKE': 'SOLID',\n",
       "     'NODE_BORDER_TRANSPARENCY': '255',\n",
       "     'NODE_BORDER_WIDTH': '2.0',\n",
       "     'NODE_CUSTOMGRAPHICS_1': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_2': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_3': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_4': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_5': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_6': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_7': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_8': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_9': 'org.cytoscape.cg.model.NullCustomGraphics,0,[ Remove Graphics ],',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_1': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_2': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_3': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_4': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_5': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_6': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_7': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_8': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_POSITION_9': 'C,C,c,0.00,0.00',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_1': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_2': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_3': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_4': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_5': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_6': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_7': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_8': '50.0',\n",
       "     'NODE_CUSTOMGRAPHICS_SIZE_9': '50.0',\n",
       "     'NODE_CUSTOMPAINT_1': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_1, name=Node Custom Paint 1)',\n",
       "     'NODE_CUSTOMPAINT_2': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_2, name=Node Custom Paint 2)',\n",
       "     'NODE_CUSTOMPAINT_3': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_3, name=Node Custom Paint 3)',\n",
       "     'NODE_CUSTOMPAINT_4': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_4, name=Node Custom Paint 4)',\n",
       "     'NODE_CUSTOMPAINT_5': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_5, name=Node Custom Paint 5)',\n",
       "     'NODE_CUSTOMPAINT_6': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_6, name=Node Custom Paint 6)',\n",
       "     'NODE_CUSTOMPAINT_7': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_7, name=Node Custom Paint 7)',\n",
       "     'NODE_CUSTOMPAINT_8': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_8, name=Node Custom Paint 8)',\n",
       "     'NODE_CUSTOMPAINT_9': 'DefaultVisualizableVisualProperty(id=NODE_CUSTOMPAINT_9, name=Node Custom Paint 9)',\n",
       "     'NODE_DEPTH': '0.0',\n",
       "     'NODE_FILL_COLOR': '#CCFFCC',\n",
       "     'NODE_HEIGHT': '40.0',\n",
       "     'NODE_LABEL_COLOR': '#000000',\n",
       "     'NODE_LABEL_FONT_FACE': 'Dialog,plain,12',\n",
       "     'NODE_LABEL_FONT_SIZE': '14',\n",
       "     'NODE_LABEL_POSITION': 'C,C,c,0.00,0.00',\n",
       "     'NODE_LABEL_ROTATION': '0.0',\n",
       "     'NODE_LABEL_TRANSPARENCY': '255',\n",
       "     'NODE_LABEL_WIDTH': '85.0',\n",
       "     'NODE_NESTED_NETWORK_IMAGE_VISIBLE': 'true',\n",
       "     'NODE_PAINT': '#FF0033',\n",
       "     'NODE_SELECTED': 'false',\n",
       "     'NODE_SELECTED_PAINT': '#FFFF00',\n",
       "     'NODE_SHAPE': 'ELLIPSE',\n",
       "     'NODE_SIZE': '35.0',\n",
       "     'NODE_TRANSPARENCY': '225',\n",
       "     'NODE_VISIBLE': 'true',\n",
       "     'NODE_WIDTH': '95.0',\n",
       "     'NODE_X_LOCATION': '0.0',\n",
       "     'NODE_Y_LOCATION': '0.0',\n",
       "     'NODE_Z_LOCATION': '0.0'},\n",
       "    'dependencies': {'nodeCustomGraphicsSizeSync': 'true',\n",
       "     'nodeSizeLocked': 'false'},\n",
       "    'mappings': {'NODE_BORDER_PAINT': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=TYPE,T=string,K=0=phenotype,V=0=#000000,K=1=smallmolecule,V=1=#FF66FF,K=2=chemical,V=2=#FF9900,K=3=stimulus,V=3=#00CCFF'},\n",
       "     'NODE_FILL_COLOR': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=TYPE,T=string,K=0=phenotype,V=0=#FFFFFF,K=1=smallmolecule,V=1=#FFCCFF,K=2=chemical,V=2=#FFCC33,K=3=stimulus,V=3=#99FFFF,K=4=complex,V=4=#00CC33'},\n",
       "     'NODE_HEIGHT': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=TYPE,T=string,K=0=phenotype,V=0=60.0,K=1=protein,V=1=30.0,K=2=proteinfamily,V=2=50.0'},\n",
       "     'NODE_LABEL': {'type': 'PASSTHROUGH', 'definition': 'COL=name,T=string'},\n",
       "     'NODE_SHAPE': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=TYPE,T=string,K=0=phenotype,V=0=RECTANGLE,K=1=smallmolecule,V=1=OCTAGON,K=2=chemical,V=2=OCTAGON,K=3=stimulus,V=3=DIAMOND,K=4=complex,V=4=ROUND_RECTANGLE'},\n",
       "     'NODE_WIDTH': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=TYPE,T=string,K=0=phenotype,V=0=110.0,K=1=complex,V=1=110.0,K=2=protein,V=2=80.0,K=3=proteinfamily,V=3=90.0'}}},\n",
       "   {'properties_of': 'edges:default',\n",
       "    'properties': {'EDGE_CURVED': 'true',\n",
       "     'EDGE_LABEL_COLOR': '#000000',\n",
       "     'EDGE_LABEL_FONT_FACE': 'Dialog,plain,10',\n",
       "     'EDGE_LABEL_FONT_SIZE': '10',\n",
       "     'EDGE_LABEL_ROTATION': '0.0',\n",
       "     'EDGE_LABEL_TRANSPARENCY': '255',\n",
       "     'EDGE_LABEL_WIDTH': '200.0',\n",
       "     'EDGE_LINE_TYPE': 'SOLID',\n",
       "     'EDGE_PAINT': '#000000',\n",
       "     'EDGE_SELECTED': 'false',\n",
       "     'EDGE_SELECTED_PAINT': '#FF0000',\n",
       "     'EDGE_SOURCE_ARROW_SELECTED_PAINT': '#FFFF00',\n",
       "     'EDGE_SOURCE_ARROW_SHAPE': 'NONE',\n",
       "     'EDGE_SOURCE_ARROW_SIZE': '6.0',\n",
       "     'EDGE_SOURCE_ARROW_UNSELECTED_PAINT': '#000000',\n",
       "     'EDGE_STACKING': 'AUTO_BEND',\n",
       "     'EDGE_STACKING_DENSITY': '0.5',\n",
       "     'EDGE_STROKE_SELECTED_PAINT': '#FFFF00',\n",
       "     'EDGE_STROKE_UNSELECTED_PAINT': '#848484',\n",
       "     'EDGE_TARGET_ARROW_SELECTED_PAINT': '#FFFF00',\n",
       "     'EDGE_TARGET_ARROW_SHAPE': 'NONE',\n",
       "     'EDGE_TARGET_ARROW_SIZE': '6.0',\n",
       "     'EDGE_TARGET_ARROW_UNSELECTED_PAINT': '#000000',\n",
       "     'EDGE_TRANSPARENCY': '255',\n",
       "     'EDGE_UNSELECTED_PAINT': '#333333',\n",
       "     'EDGE_VISIBLE': 'true',\n",
       "     'EDGE_WIDTH': '3.0',\n",
       "     'EDGE_Z_ORDER': '0.0'},\n",
       "    'dependencies': {'arrowColorMatchesEdge': 'true'},\n",
       "    'mappings': {'EDGE_LINE_TYPE': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=DIRECT,T=string,K=0=NO,V=0=EQUAL_DASH,K=1=UNK,V=1=EQUAL_DASH'},\n",
       "     'EDGE_TARGET_ARROW_SHAPE': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=DIRECT,T=string,K=0=YES,V=0=ARROW'},\n",
       "     'EDGE_UNSELECTED_PAINT': {'type': 'DISCRETE',\n",
       "      'definition': 'COL=shared interaction,T=string,K=0=down-regulates quantity,V=0=#FF3300,K=1=up-regulates quantity,V=1=#0099FF,K=2=down-regulates activity,V=2=#FF3300,K=3=down-regulates quantity by destabilization,V=3=#FF3300,K=4=down-regulates quantity by repression,V=4=#FF3300,K=5=up-regulates activity,V=5=#0099FF,K=6=up-regulates quantity by stabilization,V=6=#0099FF,K=7=up-regulates,V=7=#0099FF,K=8=down-regulates,V=8=#FF3300,K=9=up-regulates quantity by expression,V=9=#0099FF'}}}]},\n",
       " {'cartesianCoordinates': [{'cartesianLayout': [{'node': 0,\n",
       "      'x': 0.0452923119677766,\n",
       "      'y': 0.7177947654003259},\n",
       "     {'node': 1, 'x': 0.0, 'y': 0.5405120052019542},\n",
       "     {'node': 2, 'x': 0.5627443584958638, 'y': 0.6189305993357905},\n",
       "     {'node': 3, 'x': 0.6797414180240412, 'y': 0.9701690445669956},\n",
       "     {'node': 4, 'x': 0.8095146858288184, 'y': 0.8928950199409863},\n",
       "     {'node': 5, 'x': 0.5932712097211968, 'y': 0.9647951496778565},\n",
       "     {'node': 6, 'x': 0.04418698462400788, 'y': 0.3947204789015648},\n",
       "     {'node': 7, 'x': 0.04669437698793527, 'y': 0.29770501446858094},\n",
       "     {'node': 8, 'x': 0.35515550412999175, 'y': 0.9756051906402662},\n",
       "     {'node': 9, 'x': 0.8889441511628468, 'y': 0.7485409607295981},\n",
       "     {'node': 10, 'x': 0.7389193760069369, 'y': 0.8718926735301058},\n",
       "     {'node': 11, 'x': 0.7020680527151514, 'y': 0.3763929168054178},\n",
       "     {'node': 12, 'x': 0.7192027801673554, 'y': 0.08653347719497603},\n",
       "     {'node': 13, 'x': 0.8725466888351182, 'y': 0.4360872097369943},\n",
       "     {'node': 14, 'x': 0.7660360725829296, 'y': 0.23136993120215252},\n",
       "     {'node': 15, 'x': 1.0, 'y': 0.45894924678345056},\n",
       "     {'node': 16, 'x': 0.9400748432979154, 'y': 0.2520715215989529},\n",
       "     {'node': 17, 'x': 0.3000677363545341, 'y': 0.25675154393051125},\n",
       "     {'node': 18, 'x': 0.3024657971171412, 'y': 0.04284894492541376},\n",
       "     {'node': 19, 'x': 0.3768345730444027, 'y': 0.0},\n",
       "     {'node': 20, 'x': 0.49184713864009094, 'y': 0.01583827254838108},\n",
       "     {'node': 21, 'x': 0.6192398607702763, 'y': 0.007924820885611975},\n",
       "     {'node': 22, 'x': 0.4784280190562991, 'y': 0.38359934865203504},\n",
       "     {'node': 23, 'x': 0.1996087778423157, 'y': 0.5605212982093716},\n",
       "     {'node': 24, 'x': 0.3312258618648006, 'y': 0.6794342589958101}]}]},\n",
       " {'status': [{'error': '', 'success': True}]}]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Python client for the NDEx database\n",
    "import ndex2.client as nc\n",
    "\n",
    "# Unique ID for a network entry in NDEx\n",
    "uuid ='b3b352e4-b673-11eb-9e72-0ac135e8bacf'\n",
    "\n",
    "# NDEx public server URL\n",
    "ndex_url = 'http://public.ndexbio.org/'\n",
    "\n",
    "# Create an instance of NDEx client\n",
    "ndex=nc.Ndex2(ndex_url)\n",
    "print(\"starting download\")\n",
    "# Download the network in CX format\n",
    "response=ndex.get_network_as_cx_stream(uuid)\n",
    "print('Response code from NDEx: ', response.status_code)\n",
    "\n",
    "# Store the data in a Python object\n",
    "cx = response.json()\n",
    "cx"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use Cytoscape Desktop for data conversion\n",
    "\n",
    "[Cytoscape Desktop](http://manual.cytoscape.org/en/stable/Supported_Network_File_Formats.html#cytoscape-js-json) supports import/export data in CX and Cytoscape.js JSON.  Please read the document to learn more about the feature. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Render it!\n",
    "\n",
    "#### Cytoscape.js Networks\n",
    "\n",
    "By default, the widget is configured to display networks in Cytoscape.js JSON.  If you pass pass it as _*data*_ parameter, it automatically render it using default _Visual Style_ and [cose layout](http://js.cytoscape.org/#layouts/cose).  Since Cytoscape.js JSON is the default data format for this Widget, you don't need to pass _format_ option."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e6aa21cf9bde40a69753ad256c333fe6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(data={'elements': {'nodes': [{'data': {'id': 'node 1'}}, {'data': {'id': 'node 2'}}, {'data': {'id':â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Cytoscape(data=minimal_cyjs_network)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### CX Networks\n",
    "\n",
    "To display CX networks, you just need to pass the List object (CX is always a list of _aspects_ for more information about CX, please read the [official document](http://www.home.ndexbio.org/data-model/)) **AND SPECIFY FORMAT, cx**.  If the data contains Visual Style and layout, it will be applied automatically.  Otherwise, default ones will be used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e6767cc4d6bb4e4692bac94ec0c7b90d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(data=[{'numberVerification': [{'longNumber': 281474976710655}]}, {'metaData': [{'name': 'nodeAttribuâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Cytoscape(data=cx, format='cx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use files exported from Cytoscape\n",
    "\n",
    "You can easily use files edited in [Cytoscape Desktop](https://cytoscape.org/) because it supports both CX and Cytoscpae.js formats by default.  In the following example, it will display a [KEGG pathway diagram](https://www.genome.jp/kegg-bin/show_pathway?org_name=hsa&mapno=00020&mapscale=&show_description=hide) imported into Cytoscape Desktop using [KEGGScape](http://apps.cytoscape.org/apps/keggscape) and exported to Cytoscape.json files.\n",
    "\n",
    "(Network object and Visual Style are exported to separate files.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e1566153daf342ff8610d08e2ee95f23",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(data={'format_version': '1.0', 'generated_by': 'cytoscape-3.6.1', 'target_cytoscapejs_version': '~2.â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "# Network file exported from Cytoscape (in .cyjs format)\n",
    "with open('hsa00020.cyjs') as f:\n",
    "    kegg = json.load(f)\n",
    "\n",
    "# Style file exported from Cytoscape (in Cytoscape.js Style JSON format)\n",
    "with open('kegg-styles.json') as f2:\n",
    "    kegg_style = json.load(f2)\n",
    "\n",
    "# In this case, actual style is a first element in the array...\n",
    "style_obj = kegg_style[0]['style']\n",
    "    \n",
    "# Pass the style and network to the widget (layout is optional)\n",
    "Cytoscape(data=kegg, visual_style=style_obj, layout={'height': '700px'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Optional Parameters\n",
    "\n",
    "You can use most of the basic Cytoscape.js features through optional parameters.\n",
    "\n",
    "### Cell Height\n",
    "\n",
    "This widget uses [standard Layout mechanism for Jupyter Widgets](https://minrk-ipywidgets.readthedocs.io/en/latest/examples/Widget%20Styling.html#The-layout-attribute.).  To specify the cell height, you can pass it as a part of _layout_ parameter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "706c6da6a46d44f9ae486dad7a79d4e0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(data=[{'numberVerification': [{'longNumber': 281474976710655}]}, {'metaData': [{'name': 'nodeAttribuâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Cytoscape(data=cx, format='cx', layout={'height': '300px'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Automatic Graph Layout\n",
    "\n",
    "If there is no **layout_name** parameter, default layout (cose layout algorithm) will be used.  To specify one of the layout algorithms available in Cytoscape.js, just pass the name.\n",
    "\n",
    "#### Preset Layout\n",
    "\n",
    "If you have already specified (x, y) positions for nodes, the parameter will be used as the node positions.  However, if you pass this **layout_name** parameter, it will be override those original positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "cfc3a534176e477a9595bb3464b4626c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(data=[{'numberVerification': [{'longNumber': 281474976710655}]}, {'metaData': [{'name': 'nodeAttribuâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "Cytoscape(data=cx, format='cx', layout_name='grid')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Custom _Visual Style_\n",
    "\n",
    "This widget fully supports Cytoscape.js compatible styls.  Just pass your style as a dict.  For more information about styling, please visit Cytoscape.js official web site to learn how to create data mappings:\n",
    "\n",
    "* Cytoscape.js [Style](http://js.cytoscape.org/#style)\n",
    "\n",
    "Also, you can change background color by passing _background_ parameter as string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bae7f55751d24ac1957b5eb462c3d255",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(background='#000000', data=[{'numberVerification': [{'longNumber': 281474976710655}]}, {'metaData': â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Custom visual style as Python Dict\n",
    "my_style = [\n",
    "    {\n",
    "        'selector': 'node',\n",
    "        'style': {\n",
    "            'background-color': '#EFEFEF',\n",
    "            'label': 'data(name)',\n",
    "            'width': 10,\n",
    "            'height': 10,\n",
    "            'shape': 'diamond',\n",
    "            'color': '#EEEEEE',\n",
    "            'font-weight': 400,\n",
    "            'text-halign': 'right',\n",
    "            'text-valign': 'bottom',\n",
    "            'font-size': 12\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        'selector': 'edge',\n",
    "        'style': {\n",
    "            'width': 3,\n",
    "            'opacity': 0.8,\n",
    "            'line-color': 'green',\n",
    "            'target-arrow-color': 'green',\n",
    "            'target-arrow-shape': 'triangle'\n",
    "        }\n",
    "    }\n",
    "]\n",
    "\n",
    "Cytoscape(data=cx, format='cx', layout_name='breadthfirst', visual_style=my_style, background='#000000')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Use with Other Tools and Libraries\n",
    "\n",
    "This widget is a simple network visualizer, but is powerful if you use with other popular libraries.\n",
    "\n",
    "### Network X\n",
    "\n",
    "[NetworkX](https://networkx.github.io/) is a widely used library for network analysis.  It has basic visualization features, but you can use advanced visualization feature of Cytoscape.js using this widget.\n",
    "\n",
    "**NetworkX v2.x series has I/O library for Cytoscape.js objects, and we strongly recommend to use v2.2 or newer.**\n",
    "\n",
    "#### Example 1: Use NetworkX as a randome graph generator\n",
    "\n",
    "NetworkX has several randome graph generators and you can easily use the data in CyJupyter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# You must install NetworkX before running the following cells!\n",
    "import networkx as nx\n",
    "from networkx.readwrite import cytoscape_data\n",
    "from networkx.algorithms import betweenness_centrality, edge_betweenness_centrality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "252160c46c31469aba7450bb3c19ad96",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(data={'data': [], 'directed': True, 'multigraph': True, 'elements': {'nodes': [{'data': {'id': '0', â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "network1 = nx.scale_free_graph(100)\n",
    "cyjs1 = cytoscape_data(network1)\n",
    "Cytoscape(data=cyjs1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example 2: Calculate network statistics and use it in visualization\n",
    "\n",
    "NetworkX has basic visualization features, but you can replace that part with this widget."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9acd0f54aaae40eb94abf5ee88af4a6c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(background='#EEEEEE', data={'data': [], 'directed': False, 'multigraph': False, 'elements': {'nodes'â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "G = nx.waxman_graph(150)\n",
    "\n",
    "# Calculate betweenness centrarity of the graph (for both nodes and edges)\n",
    "bc = betweenness_centrality(G)\n",
    "ebc = edge_betweenness_centrality(G)\n",
    "nx.set_node_attributes(G, bc, 'betweenness')\n",
    "nx.set_edge_attributes(G, ebc, 'betweenness')\n",
    "\n",
    "# Pick min and max for each data\n",
    "bc_min = min(bc.values())\n",
    "bc_max = max(bc.values())\n",
    "ebc_min = min(ebc.values())\n",
    "ebc_max = max(ebc.values())\n",
    "\n",
    "# Create visual mappings\n",
    "\n",
    "# For nodes\n",
    "mapper = 'mapData(betweenness,' + str(bc_min) + ',' + str(bc_max) + ', 5, 50)'\n",
    "op_mapper = 'mapData(betweenness,' + str(bc_min) + ',' + str(bc_max) + ', 0.3, 1)'\n",
    "color_mapper = 'mapData(betweenness,' + str(bc_min) + ',' + str(bc_max) + ', #EFEFEF, #00897B)'\n",
    "\n",
    "# For edges\n",
    "edge_mapper = 'mapData(betweenness,' + str(ebc_min) + ',' + str(ebc_max) + ', 1, 10)'\n",
    "edge_op_mapper = 'mapData(betweenness,' + str(ebc_min) + ',' + str(ebc_max) + ', 0.3, 1)'\n",
    "\n",
    "# Build a style\n",
    "custom_visual_style = [\n",
    "    {\n",
    "        'selector': 'node',\n",
    "        'style': {\n",
    "            'background-color': color_mapper,\n",
    "            'width': mapper,\n",
    "            'height': mapper,\n",
    "            'background-opacity': op_mapper,\n",
    "            'label': \"data(id)\"\n",
    "        }\n",
    "    },\n",
    "    {\n",
    "        'selector': 'edge',\n",
    "        'style': {\n",
    "            'width': edge_mapper,\n",
    "            'opacity': edge_op_mapper,\n",
    "            'line-color': '#333333'        \n",
    "        }\n",
    "    }\n",
    "]\n",
    "\n",
    "# Convert to Cytoscape.js JSON\n",
    "cyjs2 = cytoscape_data(G)\n",
    "\n",
    "# Visualize it\n",
    "Cytoscape(data=cyjs2,  layout_name='circle', visual_style=custom_visual_style, background='#EEEEEE')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example 3: Use NetworkX as layout engine\n",
    "\n",
    "NetworkX has a nice drawing features, but all generated images are static.  If you use it as a layout engine, you can visualize it using the node positions calculated in NetworkX (or graphviz, which is supported by NetworkX):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "95910b972a8046dd87ef0362b933026d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Cytoscape(data={'data': [], 'directed': False, 'multigraph': False, 'elements': {'nodes': [{'data': {'pos': (0â€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Layout using NetworkX\n",
    "pos = nx.kamada_kawai_layout(G)\n",
    "nodes = cyjs2['elements']['nodes']\n",
    "\n",
    "SCALE_FACTOR = 700 # This adjust density of the view, because original positions are too dense...\n",
    "\n",
    "# Assign (X, Y) positions\n",
    "for node in nodes:\n",
    "    node_id = node['data']['id']\n",
    "    node['position'] = {\n",
    "        'x': pos[int(node_id)][0] * SCALE_FACTOR,\n",
    "        'y': pos[int(node_id)][1] * SCALE_FACTOR\n",
    "    }\n",
    "\n",
    "# Render it with new positions\n",
    "Cytoscape(data=cyjs2, visual_style=custom_visual_style)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
